### DISCUSSION TITLE
wasm-ld：unknown file type

### DISCUSSION THREAD

**Discussion by journey0726** [emscripten-core/emscripten#19017] [Q&A] [answered]
emcc info： 
![image](https://user-images.githubusercontent.com/73887142/226501391-1c86f2dd-317e-44ea-bd77-75dfd9f01ca6.png)
The operating system is ubantu，In the current directory，has some files: 
![image](https://user-images.githubusercontent.com/73887142/226501553-55a828a2-94e7-481a-b7cb-f63663b698a7.png)
When I use `emcc` :  `emcc -L . -ltest main.cpp -o main.html -s WASM=1 `，
Will report an error：`wasm-ld: unknow file type: ./libtest.a(libtest.so)`,
My goal is to link `main.cpp` to the `libtest.so` file, and compile it into `wasm`.
I have found many solutions, but none have been resolved。

**Comment by sbc100** [comment, marked as answer]
Somehow it looks like a file called `libtest.so` is inside of the `.a` archive `libtest.a`.

What does `llvm-ar t libtest.a` report?

If it reports `libtest.so`, what type of file is `libtest.so`?  You can use `llvm-ar -x libtest.a` to extract the files in the archive and then run `file libtest.so` to see what type of file `libtest.so` is.

**Reply by journey0726** [reply]
the `llvm-ar t libtest.a` result is `libtest.so`
the `file libtest.so` result is `libtest.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=09c425298e347ad4f003aac88ba99e358a567406, not stripped`

**Reply by journey0726** [reply]
this `.so` file was generated by `g++`.

**Reply by sbc100** [reply]
Yes that ELF file is for x86_64 not wasm.    You would need to rebuild it it with emscripten.

**Reply by journey0726** [reply]
Do you mean that I need to use `emcc` to convert the `so` file into a `wasm` file first?  And if I only have `so` file ,no source code(`libtest.cpp`),
I don't seem to have found a way to convert to a `wasm` file.

**Reply by sbc100** [reply]
If you don't have the source code and you only have and X86_64 version of the library there isn't really any way you can use it with emscripten/wasm.

**Reply by journey0726** [reply]
Do you mean other versions of the library are okay？ Which version is appropriate？ thanks.

**Reply by sbc100** [reply]
The library author would need to supply you with an emscripten version, built with emcc.

**Reply by journey0726** [reply]
the `emcc` version is `3.1.34`,the top picture has the `emcc` version.
If only the dynamic link library is available, then emcc cannot be used to compile a C++ program that uses the dynamic link library, right？

**Reply by sbc100** [reply]
Emscripten supports some forms of dynamic linking, but only with libraries compiled with emcc.     So you would need a wasm version of all your libraries dependencies.

**Reply by journey0726** [reply]
Therefore, if you only have other versions of the dynamic link library, you cannot use emcc.
I see there is a dynamic link's example using `dlopen` on the official website,does this work?

**Reply by sbc100** [reply]
Yes, but again, the thing that you are opening (the dynamic library) must be built by emcc.   

This is also true for all toolchains and platforms, you can't use dynamic libraries that are built for a different architecture.

**Reply by journey0726** [reply]
OK. I understand. Thank you very much!
